File: Examples/clustering/codes/3pt.cpp
// ========================================================================================
// Example code: how to measure the connected and reduced three-point correlation functions
// ========================================================================================

#include "ThreePointCorrelation_comoving_reduced.h"
#include "GlobalFunc.h"

// these two variables contain the name of the CosmoBolognaLib
// directory and the name of the current directory (useful when
// launching the code on remote systems)
std::string cbl::par::DirCosmo = DIRCOSMO, cbl::par::DirLoc = DIRL;


int main () {

  try {
  
    // --------------------------------------------------------------
    // ---------------- set the cosmological parameters  ------------
    // --------------------------------------------------------------

    cbl::cosmology::Cosmology cosmology {cbl::cosmology::CosmologicalModel::_Planck15_};

  
    // -----------------------------------------------------------------------------------------------------------
    // ---------------- read the input catalogue (with observed coordinates: R.A., Dec, redshift) ----------------
    // -----------------------------------------------------------------------------------------------------------
  
    std::string file_catalogue = cbl::par::DirLoc+"../input/cat.dat";

    cbl::catalogue::Catalogue catalogue {cbl::catalogue::ObjectType::_Galaxy_, cbl::CoordinateType::_observed_, {file_catalogue}, cosmology};

  
    // --------------------------------------------------------------------------------------
    // ---------------- construct the random catalogue (with cubic geometry) ----------------
    // --------------------------------------------------------------------------------------

    const double N_R = 2.; // random/data ratio
   
    cbl::catalogue::Catalogue random_catalogue {cbl::catalogue::RandomType::_createRandom_box_, catalogue, N_R};
  
    // construct the sub-regions used for jackknife and bootstrap

    std::cout << "I'm constructing the sub-regions used for jackknife and bootstrap..." << std::endl;
    const int nx = 3, ny = 3, nz = 3;
    cbl::set_ObjectRegion_SubBoxes(catalogue, random_catalogue, nx, ny, nz);
  
    // -------------------------------------------------------------------------------
    // ---------------- measure the three-point correlation functions ----------------
    // -------------------------------------------------------------------------------

    // binning parameters

    const double side_s = 5.;  // 1st side of the triangle
    const double side_u = 2.;   // ratio between the 1st and 2nd sides of the triangle (u*s)
    const int nbins = 10;        // number of bins

  
    // output data
  
    const std::string dir_output = cbl::par::DirLoc+"../output/";
    const std::string dir_triplets = dir_output;
    const std::string dir_2pt = dir_output;
    const std::string file_output = "3ptJK.dat";

  
    // measure the connected and reduced three-point correlation functions and write the output

    const auto ThreeP = cbl::measure::threept::ThreePointCorrelation::Create(cbl::measure::threept::ThreePType::_comoving_connected_, catalogue, random_catalogue, cbl::triplets::TripletType::_comoving_side_, side_s, 2., side_s*side_u, 2., nbins);

    ThreeP->measure(cbl::measure::ErrorType::_None_, dir_triplets);
  
    ThreeP->write(dir_output, file_output);

  }

  catch(cbl::glob::Exception &exc) { std::cerr << exc.what() << std::endl; exit(1); }
  
  return 0;
}

File: Examples/clustering/codes/3pt.py
# =================================================================
# Example code: how to measure the three-point correlation function
# =================================================================

# to ensure compatibility in Python versions 2.x and 3.x
from __future__ import print_function

# import Python modules for scientific computing 
import os
import numpy as np

# import the CosmoBolognaLib modules 
import CosmoBolognaLib as cbl 

# define the cosmological model, with default parameters 
cosmology = cbl.Cosmology(cbl.CosmologicalModel__Planck15_)

# Input/Output files and directories 
HOME = os.getenv("HOME")
file_catalogue = (HOME+"/CosmoBolognaLib/Examples/clustering/input/cat.dat",)
dir_output = HOME+"/CosmoBolognaLib/Examples/clustering/output/"
dir_triplets = dir_output+"triplets/"
dir_random_cat = dir_output
os.system("mkdir -p "+dir_output+" "+dir_triplets)

# read the input galaxy catalogue (with polar coordinates: RA, Dec, redshift)  
print("I'm reading the input catalogue...")
catalogue = cbl.Catalogue(cbl.ObjectType__Galaxy_, cbl.CoordinateType__observed_, file_catalogue, cosmology)

# construct the random catalogue (with cubic geometry) 
print("I'm creating the catalogue...")
N_R = 2 # random/object ratio
random_catalogue = cbl.Catalogue(cbl.RandomType__createRandom_box_, catalogue, N_R)

# construct the sub-regions used for jackknife and bootstrap
print("I'm constructing the sub-regions used for jackknife and bootstrap...")
nx = 3
ny = 3
nz = 3
cbl.set_ObjectRegion_SubBoxes(catalogue, random_catalogue, nx, ny, nz)


# measure the three-point correlation function 
  
# binnig parameters 
side_s = 20.    # 1st side of the triangle
side_u = 2.     # ratio between the 1st and 2nd sides of the triangle (u*s)
perc = 0.0225   # tolerance
nbins = 15      # number of bins

# create the object used to measure the three-point correlation function #
ThreeP = cbl.ThreePointCorrelation_comoving_reduced(catalogue, random_catalogue, cbl.TripletType__comoving_theta_, side_s, side_u, perc, nbins)

# measure the two-point correlation function 
ThreeP.measure(cbl.ErrorType__Jackknife_,dir_triplets, dir_output)

# store the output data 
file_zeta = "zeta_JK.dat"
ThreeP.write(dir_output, file_zeta, True)
file_Q_cov = "Q_cov.dat"
ThreeP.write_covariance(dir_output, file_Q_cov)
